### ISSUE TITLE
Significant & inconsistent SVG rendering in multiple Android versions

### ISSUE THREAD

**Issue by BenHenning** [oppia/oppia-android#3938]
Current issues:

| OS                   | Flavor          | SVG results (native rendering)                                            | SVG results (forced to bitmap rendering) |
|----------------------|-----------------|---------------------------------------------------------|------------------------------------------|
| KitKat (SDK 19)      | KitKat (alpha)  | Fails: SVGs don't render                                | Works                                    |
| Lollipop (SDK 22)    | KitKat (alpha)  | Fails: SVGs don't render                                | Works                                    |
| Lollipop (SDK 22)    | Lollipop+ (dev) | Fails: SVGs don't render                                | Works                                    |
| Marshmallow (SDK 23) | KitKat (alpha)  | Fails: incorrect rendering (screenspace miscalculation) | Works                                    |
| Marshmallow (SDK 23) | Lollipop+ (dev) | Fails: incorrect rendering (screenspace miscalculation) | Works                                    |
| Nougat (SDK 25)      | KitKat (alpha)  | Works                                                   | N/A                                      |
| Nougat (SDK 25)      | Lollipop+ (dev) | Works                                                   | N/A                                      |
| Oreo (SDK 27)        | KitKat (alpha)  | Works                                                   | N/A                                      |
| Oreo (SDK 27)        | Lollipop+ (dev) | Works                                                   | N/A                                      |
| Pie (SDK 28)         | KitKat (alpha)  | Works                                                   | N/A                                      |
| Pie (SDK 28)         | Lollipop+ (dev) | Works                                                   | N/A                                      |
| Q (SDK 29)           | KitKat (alpha)  | Works                                                   | N/A                                      |
| Q (SDK 29)           | Lollipop+ (dev) | Works                                                   | N/A                                      |
| R (SDK 30)           | KitKat (alpha)  | Works                                                   | N/A                                      |
| R (SDK 30)           | Lollipop+ (dev) | Works                                                   | N/A                                      |

Note that all of these uses internet-based image loading (not local caching) + production assets (SVGs for thumbnails). Also, all of the above was tested using emulators rather than physical devices.

It seems that SVG rendering starts working well at SDK 25 and up, but has (at least) two issues in earlier versions:
- Not rendering at all
- Rendering in the wrong space

These *could* have a single underlying common cause, but I actually suspect that they don't.

**Comment by BenHenning** [comment]
This also repros in #3910. I suspect I missed it since I might not have tried loading production assets in the KitKat build (though it's also possible that there's a flake here).

**Comment by BenHenning** [comment]
Aha, this might be an SVG thing (which would explain the inconsistency with thumbnails loading in developer mode since those generally rely on drawables). Some output that I saw:

```
2021-10-21 13:45:57.468 7281-7353/org.oppia.android I/dalvikvm: Could not find method android.graphics.Paint.setFontVariationSettings, referenced from method com.caverock.androidsvg.utils.SVGAndroidRenderer.selectTypefaceAndFontStyling
2021-10-21 13:45:57.468 7281-7353/org.oppia.android W/dalvikvm: VFY: unable to resolve virtual method 1344: Landroid/graphics/Paint;.setFontVariationSettings (Ljava/lang/String;)Z
2021-10-21 13:45:57.468 7281-7353/org.oppia.android D/dalvikvm: VFY: replacing opcode 0x6e at 0x0110
2021-10-21 13:45:57.468 7281-7353/org.oppia.android I/dalvikvm: Could not find method android.graphics.Paint.setFontFeatureSettings, referenced from method com.caverock.androidsvg.utils.SVGAndroidRenderer.selectTypefaceAndFontStyling
2021-10-21 13:45:57.468 7281-7353/org.oppia.android W/dalvikvm: VFY: unable to resolve virtual method 1343: Landroid/graphics/Paint;.setFontFeatureSettings (Ljava/lang/String;)V
2021-10-21 13:45:57.468 7281-7353/org.oppia.android D/dalvikvm: VFY: replacing opcode 0x6e at 0x0140
2021-10-21 13:45:57.468 7281-7353/org.oppia.android I/dalvikvm: Could not find method android.graphics.Paint.setLetterSpacing, referenced from method com.caverock.androidsvg.utils.SVGAndroidRenderer.updateStyle
2021-10-21 13:45:57.468 7281-7353/org.oppia.android W/dalvikvm: VFY: unable to resolve virtual method 1346: Landroid/graphics/Paint;.setLetterSpacing (F)V
2021-10-21 13:45:57.468 7281-7353/org.oppia.android D/dalvikvm: VFY: replacing opcode 0x6e at 0x0547
```

Interestingly, images in lessons mostly seem to work, however that might be because some are still using PNG.

**Comment by BenHenning** [comment]
Per https://github.com/oppia/androidsvg/blob/master/androidsvg/build.gradle#L6 AndroidSVG claims to have backward compatibility far before KitKat, but it seems there might be actual incompatibilities here.

https://developer.android.com/reference/android/graphics/Paint#setFontVariationSettings(java.lang.String) is an API 26 method. Oddly, I don't actually see a reference to that anywhere in AndroidSVG, so trying to figure out what's going on here.

**Comment by BenHenning** [comment]
Aha: https://github.com/BigBadaboom/androidsvg/blob/master/androidsvg/src/main/java/com/caverock/androidsvg/utils/SVGAndroidRenderer.java#L1610.

Edit: Though, this is a bit odd. Per https://github.com/oppia/androidsvg/blob/master/androidsvg/src/main/java/com/caverock/androidsvg/utils/SVGAndroidRenderer.java#L123 it shouldn't even be executing this codepath since KitKat is running. Maybe this is just a red herring error that Dalvik is printing because it's parsing bytecode that references unsupported methods, but that it doesn't actually try to execute them (since that would presumably result in a crash, instead).

My guess is that some other necessary feature is being turned off on KitKat devices. Going to try running the same build on a Lollipop emulator just to isolate the SDK version bit.

**Comment by BenHenning** [comment]
Also, I can't be sure but I think we can rule out Glide being a root cause here since images are loading fine with thumbnails (per the Glide crash that we're seeing: #3887). I'm not positive Glide isn't a factor, but it seems more likely with current evidence that this is something in AndroidSVG.

**Comment by BenHenning** [comment]
Interesting. This repros on Lollipop, so it's not KitKat-specific. Trying on Marshmallow now.

**Comment by BenHenning** [comment]
Ah, this is definitely a problem. Marshmallow loads the SVGs, but has serious issues with rendering being way off:

![image](https://user-images.githubusercontent.com/12983742/138358523-ef6ef8a8-df03-42ce-a219-0c097d5f7cd4.png)

Oddly, this seems tied to the screenspace (it seems like some coordinate translation issue is happening):

https://user-images.githubusercontent.com/12983742/138360226-c9265a91-a441-4886-9f20-95ec7cba30ae.mp4

**Comment by BenHenning** [comment]
Aha. On Marshmallow, the pre-rendered SVGs (for image blurring) actually seem to render correctly:

![image](https://user-images.githubusercontent.com/12983742/138366902-525c91b0-2549-4605-ad13-565c50a4588a.png)

In order for that to work, a large part of the image pipeline must be working:
- Loading from disk/internet
- Parsing SVG
- Converting SVG to Picture*
- Rendering Picture to drawable*

*Steps that are running partly through the Glide loading pipeline.

This actually suggests an issue with PictureDrawable, the compat library, and/or Glide (for custom drawables).

Edit: confirming that this inconsistency also exists for KitKat (& presumably Lollipop, but I didn't check), meaning it's definitely an issue with rendering pictures and not the loading pipeline.

**Comment by BenHenning** [comment]
Hmm. I'm not sure exactly why this is happening on older versions, and probably is something that would require a deeper investigation. A reasonable stop-gap might be to lock SVG-based rendering behind an SDK check and instead use the built-in bitmap conversion we already have (which might also perform better for older devices).

**Comment by BenHenning** [comment]
Confirming that bitmap rendering is a functional step-in for the missing functionality here. Given that it's a very viable workaround (and something we were considering adding as a caching/efficiency improvement in the future), I'm inclined to use it as the main solution & file TODOs to dig into the OS-specific issues in the future.

Interestingly, this discovery indicates that something very specific to rendering pictures in real-time is causing issues on older OSes, not rendering pictures at all (since the conversion to Bitmap isn't working). Note that  on Lollipop & Marshmallow devices, the Oppia splash screen also has rendering issues (but seems to be fine on KitKat and N+).

Edit: I filed #3961 to track the longer-term work here.

Edit 2: I forgot to mention that I also confirmed for all affected devices that math expressions continue to render correctly (given that they're just SVGs under the surface, but it's important to make sure they convert to bitmap correctly).
